### PR TITLE
Optimize bulk import and the load on chirpstack

### LINKED ISSUE #130: Bulk import probably times out.
We've just tested a bulk import of exactly 50 sensors.

We had to import the CSV several times to import all sensors - and technically failed to do so with one file.

The first import managed to import 29 sensors.

At first, we waited. Then we said, let there be light... and slowly a few sensors were added to the application count at a time.
First 1, then like 8, then a few refreshes more... etc.
The next day we were at 29. Which we probably reached within a few minutes, but since we were unaware of the issue at the time, we kept waiting.

So the next day we added the entire CSV again. Got the following errors on most sensors. Even some that were successfully added, so not sure what they all mean.

> SensorX | LoRaWAN | Failed | Id skal være unikt. Det pågældende Id er allerede brug
> SensorX+1 | LoRaWAN | Failed | write tcp XX.x.0.154:46986->XXX.x.0.56:5432: write: connection reset by peer
> SensorX+2 | LoRaWAN | Failed | EOF

Unique ID makes perfect sense. We are adding the same sensor several times, so this one should be there. But it wasn't consistent. 

About 4 more sensors were added during the 2nd import.

We persisted with the same CSV file, and after 5 imports we had 49/50 sensors on.

The last refused to be added through the bulk import file. No matter how many times we tried.

The error would still read a mixture between the error codes shown above, so just because there were 49 added already, it didn't just read unique ID error for all but one.

We had to manually figure out which was missing, and add it on its own.

### PR DESCRIPTION
**Pull request by AramAlsabti**
When importing a .csv file with a large number of IoT devices, timeouts are very likely to occur. The main suspect is the chirpstack API which cannot keep up as numerous calls are made to the API just for a single LoRaWAN device.

This PR refactors this flow:
- Devices are now sent in bulks of 50 instead of all at once, in separate requests.
- Devices with similar metadata are grouped together to minimize requests made to chirpstack.
- If the API's database already has the data, then requests will be made to it instead
- To preserve the individual device error messages, the backend will keep track of any errors during the process and map them to the respective devices. That way, we can still process the rest of the devices from the client

Ideally, the backend would process the CSV file. However, that would mean the user doesn't get "live" updates

[Related frontend PR](https://github.com/OS2iot/OS2IoT-frontend/pull/65) 

Fixes #130

### CODE REVIEW

File: src/services/device-management/iot-device.service.ts [hunk 1]
```diff
@@ -384,89 +482,178 @@ export class IoTDeviceService {
...
+                continue;
+            }
+
+            const deviceModelMatch = deviceModels.find(
+                model => model.id === map.iotDeviceDto.deviceModelId
             );
-        } else {
-            iotDevice.application = null;
+
+            if (deviceModelMatch?.belongsTo.id !== applicationMatch.belongsTo.id) {
```
Comment (GufCab) [inline src/services/device-management/iot-device.service.ts:579]: Verify that the DeviceModel actually exists here. Currently it throws an error about the organization not matching if an invalid DeviceModel has been chosen.

**Review by GufCab — CHANGES_REQUESTED** [review]

File: src/entities/dto/chirpstack/chirpstack-device-id.dto.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { ChirpstackDeviceContentsDto } from "./chirpstack-device-contents.dto";
+
+export type ChirpstackDeviceId = Pick<ChirpstackDeviceContentsDto, "devEUI">;
```
[the remark this answers is not in the document]
Comment (AramAlsabti) [reply src/entities/dto/chirpstack/chirpstack-device-id.dto.ts:3]: Not exactly. This is so we can do the same `x.devEUI` [check](https://github.com/OS2iot/OS2IoT-backend/pull/140/files#diff-16110d18e648539857919a3b5b1f8109e4b241d72d178311bc7aa019de51edeaR358) irregardless of whether the list came from a parameter or from chirpstack.

File: src/services/device-management/iot-device.service.ts [hunk 3]
```diff
@@ -292,29 +316,65 @@ export class IoTDeviceService {
...
     ): Promise<IoTDevice> {
-        const childType = iotDeviceTypeMap[createIoTDeviceDto.type];
-        const iotDevice = new childType();
+        const iotDevice = await this.createMany([createIoTDeviceDto], userId);
+
+        if (iotDevice[0].error) throw new BadRequestException(iotDevice[0].error);
+        return iotDevice[0].data;
+    }
+
+    async createMany(
```
[the remark this answers is not in the document]
Comment (AramAlsabti) [reply src/services/device-management/iot-device.service.ts:336]: Good point!

File: src/services/device-management/iot-device.service.ts [hunk 4]
```diff
@@ -292,29 +316,65 @@ export class IoTDeviceService {
...
+            try {
+                const childType = iotDeviceTypeMap[createIotDevice.type];
+                const iotDevice = new childType();
+                iotDevicesMaps.push({ iotDeviceDto: createIotDevice, iotDevice });
+            } catch (error) {
+                iotDevicesMaps.push({ iotDeviceDto: createIotDevice, error });
+            }
+        });
+
+        await this.mapDtoToIoTDevice(
```
[the remark this answers is not in the document]
Comment (AramAlsabti) [reply src/services/device-management/iot-device.service.ts:353]: I agree. The issue is that a device can be invalidated in any one of the operations which is why invalid devices are filtered out before the next operation. Any suggestions?

[the remark this answers is not in the document]
Comment (AramAlsabti) [reply src/services/device-management/iot-device.service.ts:353]: Good point. I added as documentation to the method

File: src/entities/dto/chirpstack/chirpstack-device-id.dto.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (GufCab) [reply src/entities/dto/chirpstack/chirpstack-device-id.dto.ts:3]: Neat

File: src/services/device-management/iot-device.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (AramAlsabti) [reply src/services/device-management/iot-device.service.ts]: Alternatively, each method could return a list of valid devices. The input of the next method will be that list. Each method would then have to keep track of valid/invalid devices. I'm not sure if that's just as convoluted as the above, so I'll go with adding a comment at each step

**Review by GufCab — APPROVED** [review]

**Review by GufCab — APPROVED** [review]

**Pull request merged**